csharp/src/Drivers/Apache/Hive2/HiveServer2Statement.cs

```diff
@@ -587,5 +590,137 @@ protected internal QueryResult EnhanceGetColumnsResult(Schema originalSchema, IR
 
             return new QueryResult(rowCount, new HiveServer2Connection.HiveInfoArrowStream(enhancedSchema, enhancedData));
         }
+
+        private async Task<QueryResult> GetColumnsExtendedAsync(CancellationToken cancellationToken = default)
+        {
+            // 1. Get all three results at once
+            var columnsResult = await GetColumnsAsync(cancellationToken);
+            if (columnsResult.Stream == null) return columnsResult;
+
+            var pkResult = await GetPrimaryKeysAsync(cancellationToken);
+
+            // For FK lookup, we need to pass in the current catalog/schema/table as the foreign table
+            var resp = await Connection.GetCrossReferenceAsync(
+                null,
+                null,
+                null,
+                CatalogName,
+                SchemaName,
+                TableName,
+                cancellationToken);
+
+            var fkResult = await GetQueryResult(resp.DirectResults, cancellationToken);
+
+
+            // 2. Read all batches into memory
+            RecordBatch? columnsBatch = null;
+            RecordBatch? pkBatch = null;
+            RecordBatch? fkBatch = null;
+
+            // Extract column data
+            using (var stream = columnsResult.Stream)
+            {
+                columnsBatch = await stream.ReadNextRecordBatchAsync(cancellationToken);
```

Comment:
So we're guaranteed, then, that all the results will be in a single `RecordBatch`? Consider adding a comment to that effect.